Sliders.py

```diff
@@ -0,0 +1,367 @@
+        self.cbar = self.figure.colorbar(self.im)
+
+        # Label the axes
+        self.label_axes()
+
+    def change_scale(self, scale):
+
+        self.curr_scale = scale
+
+        try:
+            self.im.set_norm(self.norms[scale])
+            self.update_colourbar()
+        except:
+            self.ax.set_yscale(scale)
+
+        self.canvas.draw()
+
+    def update_colourbar(self):
+
+        self.cbar.remove()
+        self.cbar = self.figure.colorbar(self.im)
+
+    def press_button(self, dim, curr_axis_no, neighb_axis_no):
+
+        def slice_changer():
+
+           # Ignore the instruction if another dimension has already been selected for the same axis
+            if self.axis_already_selected(dim,curr_axis_no):
+                dim.buttons[curr_axis_no].setChecked(False)
+                return
+
+            # Set this dimension as an axes
+            if dim.buttons[curr_axis_no].isChecked():
+
+                # Disable the matching slider and stepper
+                dim.slider.setVisible(False)
+                dim.stepper.setVisible(False)
+
+                # Switch off the neighbouring button
+                dim.buttons[neighb_axis_no].setChecked(False)
+
+                # Set current axis to this dimension
+                self.axes[curr_axis_no] = dim
+
+                # Remvoe this dimension from the slice dictionary
+                self.slice_selection.pop(dim.name, None)
+
+            # Unset this dimension as an axis
+            else:
+
+                # Enable the matching slider and stepper
+                dim.slider.setVisible(True)
+                dim.stepper.setVisible(True)
+
+                # Set current axis to None
+                self.axes[curr_axis_no] = None
+
+                # Place this dimension in the slice dictionary
+                self.slice_selection[dim.name] = 0
+
+            # Change the view if both a X and a Y axis have been selected
+            self.change_view()
+
+        return slice_changer
+
+    def axis_already_selected(self,curr_dim,n_curr_axis):
+
+        # Check that the same axis has already been selected elsewhere (i.e., prevent to difference X buttons from being checked at the same time)
+        for dim in self.dims:
+            if dim.buttons[n_curr_axis].isChecked() and dim != curr_dim:
+                return True
+        return False
+
+    def stepper_changer_creator(self,dim):
+
+        # Sequence to be carried out when the stepper detects a change
+        def stepper_changer():
+
+            # Obtain the stepper value
+            stepper_val = dim.stepper.value()
+
+            # Change slider value to match the stepper
+            dim.slider.setValue(stepper_val)
+
+            # Update the slice selection dictionary
+            self.slice_selection[dim.name] = stepper_val
+
+            # Change the view
+            self.change_view()
+
+        return stepper_changer
+
+    def slider_changer_creator(self,dim):
```

Comment:
The functions that gets passed to `.connect()` for the slider/buttons needs to behave differently for each dimension. I couldn't see a way of doing this with lambdas or within the Dimension class because it seems to rely on knowledge of what the other Dimensions are doing as well. Does that make sense?

Reply:
I am not sure I understand what you are doing exactly. Why do you not use the slider value that is sent with the `valueChanged` signal? Wouldn't that simplify the mechanism?